examples/ringbuffer/main.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/ringbuf"
+	"golang.org/x/sys/unix"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc clang-11 RingBufferExample ./bpf/ringbuffer_example.c -- -nostdinc -I../headers
+
+// An Event represents a ringbuf event sent to userspace from the eBPF program
+// running in the kernel. Note that this must match the C event_t structure,
+// and that both C and Go structs must be aligned same way.
+type Event struct {
+	PID  uint32
+	Comm [80]byte
+}
+
+func main() {
+	// Name of the kernel function to trace.
+	fn := "sys_execve"
+
+	// Subscribe to signals for terminating the program.
+	stopper := make(chan os.Signal, 1)
+	signal.Notify(stopper, os.Interrupt, syscall.SIGTERM)
+
+	// Allow the current process to lock memory for eBPF resources.
+	rrl, err := ebpf.RemoveMemlockRlimit()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Load pre-compiled programs and maps into the kernel.
+	objs := RingBufferExampleObjects{}
+	if err := LoadRingBufferExampleObjects(&objs, nil); err != nil {
+		log.Fatalf("loading objects: %v", err)
+	}
+	defer objs.Close()
+
+	// Revert the process' rlimit after eBPF resources have been loaded.
+	if err := rrl(); err != nil {
+		log.Fatal(err)
+	}
+
+	// Open a Kprobe at the entry point of the kernel function and attach the
+	// pre-compiled program. Each time the kernel function enters, the program
+	// will emit an event containing pid and command of the execved task.
+	kp, err := link.Kprobe(fn, objs.KprobeExecve)
+	if err != nil {
+		log.Fatalf("opening kprobe: %s", err)
+	}
+	defer kp.Close()
+
+	// Open a ringbuf reader from userspace RINGBUF map described in the
+	// eBPF C program.
+	rd, err := ringbuf.NewReader(objs.Events)
+	if err != nil {
+		log.Fatalf("opening ringbuf reader: %s", err)
+	}
+	defer rd.Close()
+
+	// Close the reader when the process receives a signal, which will exit
+	// the read loop.
+	go func() {
+		<-stopper
+
+		if err := rd.Close(); err != nil {
+			log.Fatalf("closing ringbuf reader: %s", err)
+		}
+	}()
+
+	log.Println("Waiting for events..")
+
+	for {
+		record, err := rd.Read()
+		if err != nil {
+			if errors.Is(err, ringbuf.ErrClosed) {
+				log.Println("Received signal, exiting..")
+				return
+			}
+			log.Printf("reading from reader: %s", err)
```

Comment:
Sorry, coming back to this: why not abort the program here and in the tracepoint example? And why not `continue` after logging the error if we want to keep going?

Reply:
Don't know exactly all the possible errors that may arise, I think it'll be fine if we just print the error and then let the user terminate the program, if needed.
I agree that we should add a `continue`, since it would be pointless to try to extract an event from the ring buffer in case of error. I'll fix this in both examples.

Reply:
Actually, I see that the uretprobe example has the same issue, I'll fix that, too.